### ISSUE TITLE
provide type support for .eslintrc.js

### ISSUE THREAD

**Issue by tadhgmister** [typescript-eslint/typescript-eslint#2488]
I really appreciate my IDE's suggestions and intellisense, and it always bugged me that the `eslintrc` had no support.

Today I found that the basic structure of the `eslintrc` format is present [in `Linter.ts`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/experimental-utils/src/ts-eslint/Linter.ts#L175) so today I added this to my `.eslintrc.js` for my personal repo:

```js
// @ts-check
/** @type {import("@typescript-eslint/experimental-utils/dist/ts-eslint/Linter").Linter.Config} */
const config = {
     ...
};
module.exports = config;
```

This internal type doesn't contain any rule information but I think an official version of this would be desirable:
- it could be generated with a similar script to [`generate-configs.ts`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/tools/generate-configs.ts) 
- both type check, and give auto-complete for specifying options
- can make configuration for base rules that don't work so users get errors like this:

![no-shadow error with doc-string telling them to use the correct rule][1]

([code that produces above message](https://github.com/tadhgmister/typescript-eslint/commit/bdc40b48554ca61a7e08e12106d44dbdf937ab64))

As well with namespace merging it should be possible to provide an easy way for other packages to define a `.d.ts` file to define their own rule options, something like this would be useful:

```ts
declare namespace ESLINT {
    type Severity = 0 | 1 | 2 | "off" | "warn" | "error";
    type RuleOption<Settings extends unknown[] = []> = Severity | [Severity, ...Settings];
    interface Rules extends Record<string, RuleOption<unknown[]>> {
        "no-shadow": "off";
        "@typescript-eslint/no-shadow": RuleOption<[]>;
    }
}
// in another library:
declare namespace ESLINT {
    interface Rules {
        "import/first": RuleOption; // specific options specified
    }
}
```
I think there might be a way to allow this without polluting the global namespace but it'd require some investigation.

In any case I do think this is worth some consideration and I am more than happy to implement It.

[1]: https://i.stack.imgur.com/ML6CL.png

**Comment by tadhgmister** [comment]
never mind, duplicate of #1296,
